torchao/prototype/inductor/int8_sdpa_lowering.py

```diff
@@ -0,0 +1,140 @@
+
+op_int8_sdpa = ExternKernelChoice(
+    torch.ops.torchao.scaled_dot_product_int8.default,
+    "torchao::scaled_dot_product_int8",
+    has_out_variant=False,
+    use_fallback_kernel=True,
+    op_overload=torch.ops.torchao.scaled_dot_product_int8.default,
+)
+
+
+def register_int8_sdpa():
+    @register_lowering(
+        torch.ops.torchao.scaled_dot_product_int8.default, type_promotion_kind=None
+    )
+    def int8_sdpa(
+        query: TensorBox,
+        key: TensorBox,
+        value: TensorBox,
+        attn_mask: Optional[TensorBox],
+        dropout: float,
+        is_causal: bool,
+        scale: Optional[float] = None,
+        q_scale: Optional[float] = 1.0,
+        q_zp: Optional[int] = 0,
+        k_scale: Optional[float] = 1.0,
+        k_zp: Optional[int] = 0,
+        v_scale: Optional[float] = 1.0,
+        v_zp: Optional[int] = 0,
+        a_scale: Optional[float] = 1.0,
+        a_zp: Optional[int] = 0,
+        o_scale: Optional[float] = 1.0,
+        o_zp: Optional[int] = 0,
+    ) -> TensorBox:
+        choices: list[ChoiceCaller] = []
+
+        (
+            query,
+            key,
+            value,
+            attn_mask,
+        ) = maybe_realize(
+            [
+                query,
+                key,
+                value,
+                attn_mask,
+            ]
+        )
+
+        if (
+            query.get_dtype() is not torch.uint8
+            or key.get_dtype() is not torch.uint8
+            or value.get_dtype() is not torch.uint8
+        ):
+            raise NotImplementedError(
+                "Only `torch.uint8` is supported in Int8 SDPA template for CPU device. "
+                f"Found input tensors are `{query.get_dtype()}`,`{key.get_dtype()}`,`{value.get_dtype()}`."
+            )
+
+        # Construct output layout with strides matching the query.
+        out_size = query.get_size()
+        fill_order = get_fill_order(query.get_stride())
+        out_strides = construct_strides(out_size, fill_order)
+
+        layout = FixedLayout(
+            query.get_device(),
+            query.get_dtype(),
+            out_size,
+            stride=[sympy.sympify(s) for s in out_strides],
+        )
+        input_nodes = [query, key, value]
+        if attn_mask is not None:
+            input_nodes.append(attn_mask)
+
+        # use template if machine has amx
+        if torch._C._cpu._is_amx_tile_supported():
+            CppInt8SdpaTemplate.add_choices(
+                choices=choices,
+                input_nodes=input_nodes,
+                layout=layout,
+                scale=scale,
+                q_scale=q_scale,
+                q_zp=q_zp,
+                k_scale=k_scale,
+                k_zp=k_zp,
+                v_scale=v_scale,
+                v_zp=v_zp,
+                a_scale=a_scale,
+                a_zp=a_zp,
+                o_scale=o_scale,
+                o_zp=o_zp,
+            )
+
+        if len(choices) == 0:
```

Comment:
Just to confirm, does that mean we prefer using template based solution when available instead of tuning?

Reply:
Thanks, you are absolutely right. The two algorithms are almost the same now. In the future, the template one is expected to have a better perf.